plugins/RocketsPlugin/RocketsPlugin.cpp

```diff
@@ -345,20 +345,97 @@ class RocketsPlugin::Impl
 #endif
     }
 
-    template <typename T>
-    void _rebroadcast(const std::string& endpoint, const T& obj,
-                      const uintptr_t clientID)
+    void _rebroadcast(const std::string& endpoint,
+                      const rockets::ws::Request& request)
     {
-        _delayedNotify([&, json = to_json(obj) ] {
+        _delayedNotify([&, request] {
             if (_rocketsServer->getConnectionCount() > 1)
             {
                 const auto& msg =
-                    rockets::jsonrpc::makeNotification(endpoint, json);
-                _rocketsServer->broadcastText(msg, {clientID});
+                    rockets::jsonrpc::makeNotification(endpoint,
+                                                       request.message);
+                _rocketsServer->broadcastText(msg, {request.clientID});
             }
         });
     }
 
+    // Utilty to change current client while we are handling a message to skip
+    // notification to the current client and to trigger a delayed notify to
+    // avoid a deadlock which happens when sending a message from within a
+    // message handler.
+    struct ScopedCurrentClient
+    {
+        ScopedCurrentClient(uintptr_t& currentClientID, const uintptr_t newID)
+            : _currentClientID(&currentClientID)
+        {
+            *_currentClientID = newID;
+        }
+
+        ~ScopedCurrentClient() { *_currentClientID = NO_CURRENT_CLIENT; }
+    private:
+        uintptr_t* _currentClientID;
+    };
+
+    void _bindEndpoint(const std::string& method,
+                       rockets::jsonrpc::ResponseCallback action)
+    {
+        _jsonrpcServer->bind(method, [&, action](
+                                         rockets::jsonrpc::Request request) {
+            ScopedCurrentClient scope(_currentClientID, request.clientID);
+            return action(request);
+        });
+    }
+
+    template <typename Params, typename RetVal>
+    void _bindEndpoint(const std::string& method,
+                       std::function<RetVal(Params)> action)
+    {
+        _jsonrpcServer->bind(method, [&, action](
+                                         rockets::jsonrpc::Request request) {
+            ScopedCurrentClient scope(_currentClientID, request.clientID);
+
+            Params params;
+            if (!::from_json(params, request.message))
+                return Response::invalidParams();
+            try
+            {
+                const auto& ret = action(std::move(params));
+                return Response{to_json(ret)};
+            }
+            catch (const rockets::jsonrpc::response_error& e)
+            {
+                return Response{Response::Error{e.what(), e.code}};
+            }
+        });
+    }
+
+    template <typename Params>
+    void _connectEndpoint(const std::string& method,
```

Comment:
Why these methods are not called bind as well? Because they return OK?

Reply:
Because in rockets they are called connect as well.